osu.Game/Beatmaps/FramedBeatmapClock.cs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Diagnostics;
+using osu.Framework;
+using osu.Framework.Allocation;
+using osu.Framework.Audio.Track;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Timing;
+using osu.Game.Configuration;
+using osu.Game.Database;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Beatmaps
+{
+    /// <summary>
+    /// A clock intended to be the single source-of-truth for beatmap timing.
+    ///
+    /// It provides some functionality:
+    ///  - Applies (and tracks changes of) beatmap, user, and platform offsets.
+    ///  - Adjusts <see cref="Seek"/> operations to account for said offsets, seeking in raw time values.
+    ///  - Exposes track length.
+    ///  - Allows changing the source to a new track (for cases like editor track updating).
+    /// </summary>
+    public class FramedBeatmapClock : Component, IFrameBasedClock, IAdjustableClock, ISourceChangeableClock
+    {
+        private readonly bool applyOffsets;
+
+        /// <summary>
+        /// The length of the underlying beatmap track. Will default to 60 seconds if unavailable.
+        /// </summary>
+        public double TrackLength => Track?.Length ?? 60000;
+
+        /// <summary>
+        /// The underlying beatmap track, if available.
+        /// </summary>
+        public Track? Track { get; private set; } // TODO: virtual rather than null?
```

Comment:
Was referring to having a placeholder `TrackVirtual` 😅 . Wasn't meant to stay as a TODO though. Applying is quite easy, so I'll do so if it passes tests:

```diff
diff --git a/osu.Game/Beatmaps/FramedBeatmapClock.cs b/osu.Game/Beatmaps/FramedBeatmapClock.cs
index 90ff0588a6..f1d02b7894 100644
--- a/osu.Game/Beatmaps/FramedBeatmapClock.cs
+++ b/osu.Game/Beatmaps/FramedBeatmapClock.cs
@@ -31,12 +31,12 @@ public class FramedBeatmapClock : Component, IFrameBasedClock, IAdjustableClock,
         /// <summary>
         /// The length of the underlying beatmap track. Will default to 60 seconds if unavailable.
         /// </summary>
-        public double TrackLength => Track?.Length ?? 60000;
+        public double TrackLength => Track.Length;
 
         /// <summary>
         /// The underlying beatmap track, if available.
         /// </summary>
-        public Track? Track { get; private set; } // TODO: virtual rather than null?
+        public Track Track { get; private set; } = new TrackVirtual(60000);
 
         /// <summary>
         /// The total frequency adjustment from pause transforms. Should eventually be handled in a better way.
@@ -144,7 +144,7 @@ private double totalAppliedOffset
 
         public void ChangeSource(IClock? source)
         {
-            Track = source as Track;
+            Track = source as Track ?? new TrackVirtual(60000);
             decoupledClock.ChangeSource(source);
         }
 

```